File: Algorithms/src/hard/Previous Coprime Element.swift
import Foundation

// MARK: - Fast Input
final class FastScanner {
    private var data: [UInt8] = Array(FileHandle.standardInput.readDataToEndOfFile()) + [0]
    private var idx: Int = 0

    func readInt() -> Int {
        var num = 0
        var sign = 1
        while data[idx] == 10 || data[idx] == 13 || data[idx] == 32 {
            idx += 1
        }
        if data[idx] == 45 {
            sign = -1
            idx += 1
        }
        while data[idx] >= 48 {
            num = num * 10 + Int(data[idx] - 48)
            idx += 1
        }
        return num * sign
    }
}

// MARK: - Globals
let MAXM = 100_002
var primes = Array(repeating: 0, count: MAXM)
var a: [Int] = []

func gcd(_ x: Int, _ y: Int) -> Int {
    var a = x, b = y
    while b != 0 {
        let t = a % b
        a = b
        b = t
    }
    return a
}

// MARK: - Sieve Precomputation
func precomp() {
    for i in 0..<MAXM {
        primes[i] = i
    }
    for i in 2..<MAXM {
        if primes[i] == i {
            var j = i * 2
            while j < MAXM {
                primes[j] = i
                j += i
            }
        }
    }
}

// MARK: - Unique Prime Factors (Inclusion–Exclusion)
func uniqueFactors(_ x: Int) -> [(Int, Int)] {
    var used: [Int: Bool] = [:]
    var factors: [Int] = []
    var v = x

    while v > 1 {
        let p = primes[v]
        if used[p] == nil {
            used[p] = true
            factors.append(p)
        }
        v /= p
    }

    let m = 1 << factors.count
    var result: [(Int, Int)] = []

    for mask in 1..<m {
        var prod = 1
        var bits = 0
        for i in 0..<factors.count {
            if (mask & (1 << i)) != 0 {
                prod *= factors[i]
                bits += 1
            }
        }
        result.append((prod, bits & 1 == 1 ? 1 : -1))
    }
    return result
}

// MARK: - Segment Tree Node
final class Items {
    var comps: Int = 0
    var ok: [Int: Int] = [:]

    func initVal(_ val: Int) {
        comps = 1
        ok.removeAll()
        for f in uniqueFactors(val) {
            ok[f.0, default: 0] += 1
        }
    }

    func check(_ val: Int) -> Bool {
        var cnt = comps
        for f in uniqueFactors(val) {
            cnt -= (ok[f.0] ?? 0) * f.1
        }
        return cnt > 0
    }
}

// MARK: - Segment Tree
final class SegTree {
    var nodes: [Items]
    var arr: [Int]

    init(_ input: [Int]) {
        self.arr = input
        self.nodes = Array(repeating: Items(), count: 4 * input.count + 5)
        build(0, input.count, 1)
    }

    func build(_ l: Int, _ r: Int, _ k: Int) {
        if l + 1 == r {
            nodes[k].initVal(arr[l])
            return
        }
        let mid = (l + r) >> 1
        build(l, mid, k << 1)
        build(mid, r, k << 1 | 1)

        for (key, val) in nodes[k << 1].ok {
            nodes[k].ok[key, default: 0] += val
        }
        for (key, val) in nodes[k << 1 | 1].ok {
            nodes[k].ok[key, default: 0] += val
        }
        nodes[k].comps = nodes[k << 1].comps + nodes[k << 1 | 1].comps
    }

    func query(_ l: Int, _ r: Int, _ k: Int, _ pos: Int) -> Int {
        if r - l == 1 {
            if l >= pos { return -1 }
            if gcd(a[pos], a[l]) != 1 { return -1 }
            return l
        }
        let mid = (l + r) >> 1

        if mid < pos && nodes[k << 1 | 1].check(arr[pos]) {
            let res = query(mid, r, k << 1 | 1, pos)
            if res != -1 { return res }
        }
        return query(l, mid, k << 1, pos)
    }
}

// MARK: - Solve
func solve(_ scanner: FastScanner) {
    let n = scanner.readInt()
    a = Array(repeating: 0, count: n)
    for i in 0..<n {
        a[i] = scanner.readInt()
    }

    let tree = SegTree(a)
    var output = ""
    for i in 0..<n {
        output += "\(tree.query(0, n, 1, i)) "
    }
    print(output)
}

// MARK: - Main
let scanner = FastScanner()
precomp()
let t = scanner.readInt()
for _ in 0..<t {
    solve(scanner)
}

File: Algorithms/src/hard/Previous Coprime Element.cpp
#include<bits/stdc++.h>
// #include <ext/pb_ds/assoc_container.hpp> 
// #include <ext/pb_ds/tree_policy.hpp> 
using namespace std;
// using namespace __gnu_pbds;
#define int long long int
#define ld long double
#define F first
#define S second
#define P pair<int,int>
#define V vector
#define eb emplace_back
#define pb push_back
#define mp make_pair
#define endl "\n"
#define all(x) x.begin(), x.end()
#define forstl(i, s)  for (__typeof ((s).end ()) i = (s).begin (); i != (s).end (); ++i)
// #pragma comment(linker, '/STACK:200000')        //prevent stack overflow ...STACK Limiter
// #pragma comment(linker, "/stack:200000000")
// #pragma GCC optimize("Ofast")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx,tune=native")
#define time() cerr << "Time : " << (double)clock() / (double)CLOCKS_PER_SEC << "s\n" 
#define db(...) __f(#__VA_ARGS__, __VA_ARGS__)
const int MOD = 1e9 + 7;
const int MOD2 = 998244353;
ld PI=2*acos(0.0);
 
template <typename T> T gcd(T a, T b){return (b==0)?a:gcd(b,a%b);}
template <typename T> T lcm(T a, T b){return a*(b/gcd(a,b));}   
template <typename T> T mod_exp(T b, T p, T m){T x = 1;while(p){if(p&1)x=(x*b)%m;b=(b*b)%m;p=p>>1;}return x;}
template <typename T> T invFermat(T a, T p){return mod_exp(a, p-2, p);}
template <typename T> T exp(T b, T p){T x = 1;while(p){if(p&1)x=(x*b);b=(b*b);p=p>>1;}return x;}
// typedef tree<P, null_type, less<P>, rb_tree_tag, tree_order_statistics_node_update> ordered_multiset;
string tostring(int number){stringstream ss; ss<<number; return ss.str();}
int toint(const string &s) {stringstream ss; ss << s; int x; ss >> x; return x;}
 
template <typename Arg1>
void __f(const char* name, Arg1&& arg1) { cout << name << " : " << arg1 << '\n'; }
template <typename Arg1, typename... Args>
void __f(const char* names, Arg1&& arg1, Args&&... args) {
    const char* comma = strchr(names + 1, ',');
    cout.write(names, comma - names) << " : " << arg1 << " | "; __f(comma + 1, args...);
}   

const int maxm = 1e5+2;
vector<int>primes(maxm);
map<int,int>empty_map;
vector<int>a;

void precomp()
{   
    for(int i=0;i<maxm;i++){
        primes[i] = i;
    }
    for(int i=2;i<maxm;i++){
        if(primes[i] == i){
            for(int j=2*i;j<maxm;j+=i){
                primes[j] = i;
            }
        }
    }
}

vector<P>uniqueFactors(int x)
{
    map<int,int>visFacts;
    int ok = x;
    vector<int>facts;
    while(ok > 1){
        if(!visFacts[primes[ok]]){
            visFacts[primes[ok]] = true;
            facts.pb(primes[ok]);
        }
        ok /= primes[ok];
    }
    int me = (1LL<<(facts.size()));
    vector<P>allProducts;
    for(int j=1;j<me;j++){
        int bits = 0;
        int prod = 1;
        for(int k=0;k<facts.size();k++){
            if(j&(1LL<<k)){
                prod *= facts[k];
                bits++;
            }
        }
        if(bits&1){
            allProducts.pb({prod,1});
        }
        else{
            allProducts.pb({prod,-1});
        }
    }
    return allProducts;
}

class items {
public:
    int comps;
    map<int, int> ok;

    items() : comps(0) {}

    void init(int val) {
        comps = 1;
        for (auto& factor : uniqueFactors(val)) {
            ok[factor.F] += 1;
        }
    }

    bool check(int val) {
        int count = comps;
        for (auto& x : uniqueFactors(val)) {
            count -= ok[x.F]*x.S;
        }
        return count > 0;
    }
};

class segtree {
public:
    vector<items> nodes;
    vector<int> arr;

    segtree(const vector<int>& input) : arr(input) {
        nodes.resize(4 * input.size());
        build(0, input.size(), 1);
    }

    void build(int x, int y, int k) {
        if (x + 1 == y) {
            nodes[k].init(arr[x]);
            return;
        }
        int mid = (x + y) / 2;
        build(x, mid, 2 * k);
        build(mid, y, 2 * k + 1);

        for (const auto& x : nodes[2 * k].ok) {
            nodes[k].ok[x.F] += x.S;
        }
        for (const auto& x : nodes[2 * k + 1].ok) {
            nodes[k].ok[x.F] += x.S;
        }

        nodes[k].comps = nodes[2 * k].comps + nodes[2 * k + 1].comps;
    }

    int query(int x, int y, int k, int pos) {
        if ((y-x) == 1) {
            if(x >= pos){
                return -1;
            }
            if(gcd(a[pos],a[x])!=1){
                return -1;
            }
            return x;
        }
        int mid = (x + y) / 2;
        if (mid < pos and nodes[2 * k + 1].check(arr[pos])) {
            int curr = query(mid, y, 2 * k + 1, pos);
            if (curr != -1) {
                return curr;
            }
        }
        return query(x, mid, 2 * k, pos);
    }
};

void solve()
{   
    int n; cin>>n;
    a.clear();
    a.resize(n);
    for(auto &x : a){
        cin>>x;
    }
    segtree tree(a);
    for (int i = 0; i < a.size(); ++i) {
        cout << tree.query(0, n, 1, i) << " ";
    }
    cout<<endl;
}

int32_t main()
{
    ios_base:: sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    #ifndef ONLINE_JUDGE                    //file start
    freopen("input.txt","r",stdin);
    freopen("output.txt","w",stdout);   
    #endif                                  //file end
    // int cases=1;
    precomp();
    int t; cin>>t;
    while(t--)
    solve();
    return 0;
}
 
// 1LL check ?
